server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java

```diff
@@ -119,6 +119,102 @@ interface VectorByteDataConsumer {
         void accept(byte value, int index);
     }
 
+    public static class VectorData {
+
+        private final float[] floatVector;
+        private final byte[] byteVector;
+
+        private VectorData(float[] floatVector) {
+            this.floatVector = Objects.requireNonNull(floatVector, "the provided float vector cannot be null");
+            this.byteVector = null;
+        }
+
+        private VectorData(byte[] byteVector) {
+            this.floatVector = null;
+            this.byteVector = Objects.requireNonNull(byteVector, "the provided byte vector cannot be null");
+        }
+
+        public boolean isFloatVector() {
+            return floatVector != null;
+        }
+
+        public boolean isByteVector() {
+            return byteVector != null;
+        }
+
+        public byte[] asByteVector() {
+            if (isByteVector()) {
+                return byteVector;
+            } else if (isFloatVector()) {
+                ElementType.BYTE.checkVectorBounds(floatVector);
+                byte[] vec = new byte[floatVector.length];
+                for (int i = 0; i < floatVector.length; i++) {
+                    vec[i] = (byte) floatVector[i];
+                }
+                return vec;
+            }
+            return new byte[0];
+        }
```

Comment:
It is nice to have this for the mapper. However, this should have checks to ensure that if this is called, its actually a byte vector (meaning, whole numbers between Byte.MIN_VALUE|MAX_VALUE)

Reply:
Pretty sure we iterate twice already. Though, it doesn't make sense to "bounds check" once the vector has switched to `byte[]` already as you know for sure that `byte[]` values are whole and between min/max ;) (as they are `byte` values).

It seems like `VectorData` should have an `as...` method and a `asElementType(ElementType)` to handle these weird scenarios.

Reply:
The `asByteVector` and `asFloatVector` methods are currently called only when moving to explicit implementations (e.g. `DenseVectorFieldType#createKnnByteQuery` and `DenseVectorFieldType#createKnnFloatQuery`), and in the `toXContent` methods.  So, in most cases, while the underlying vector has been converted to byte, the `VectorData` record itself is no longer used, so we won't need to re-transform the data. 

Would it make sense to have a "rewrite"-like method to return a new `VectorData` with `byte[]` instead of `float[]`? 

> It seems like VectorData should have an `as...` method and a `asElementType(ElementType)` to handle these weird scenarios.

Not sure the distinction of the two usages is clear to me.  Could you please provide an example of a scenario that the `asElementType` method would handle to help me understand the intent?

Reply:
> Not sure the distinction of the two usages is clear to me. Could you please provide an example of a scenario that the asElementType method would handle to help me understand the intent

Maybe I misunderstand then. I was thinking a query:

 - Starting from an old node (parsed as `float[]`)
 - Serialized to new node (read into `VectorData`)
 - Toquery is called but really its a `element_type: byte` so we have to check dimensions and transform into `byte[]` to create the correct Lucene query.

Reply:
Yeah, the process remains pretty much the same - exactly as you described it. There was also a `parseFloat` method that tried to eagerly read into `byte[]` during initial parsing (from either XContent or older nodes) but has now been removed to better distinguish between explicit byte & float vectors and fail for hex-float combination.

So now, hex aside, we pass around `VectorData` records holding float vectors (this hasn't changed)  and call the `asFloatVector` / `asByteVector` in `DenseVectorFieldMapper` in the `createKnnQuery` and `createExactKnnQuery` methods, depending on the element type. At that point, as you've mentioned we do the dimension check / bound check etc and pass the `byte[]` instance from now on. 

I might have misunderstood/missing something , but AFAICT the dimensionality check and conversion happens only at that point, hence why it isn't clear to me the need for an additional `asElementType` method.  

Also, please note that I've just pushed a new set of changes moving `VectorData` outside of `DenseVectorFieldMapper` and taking care of serialization as suggested.